File: hadrian/src/main/scala/com/opendatagroup/hadrian/lib1/model/tree.scala
// Copyright (C) 2014  Open Data ("Open Data" refers to
// one or more of the following companies: Open Data Partners LLC,
// Open Data Research LLC, or Open Data Capital LLC.)
// 
// This file is part of Hadrian.
// 
// Licensed under the Hadrian Personal Use and Evaluation License (PUEL);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://raw.githubusercontent.com/opendatagroup/hadrian/master/LICENSE
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.opendatagroup.hadrian.lib1.model

import scala.annotation.tailrec
import scala.collection.immutable.ListMap
import scala.collection.JavaConversions._

import org.apache.avro.AvroRuntimeException
import org.apache.avro.SchemaCompatibility.checkReaderWriterCompatibility
import org.apache.avro.SchemaCompatibility.SchemaCompatibilityType
import org.apache.avro.Schema

import com.opendatagroup.hadrian.ast.LibFcn
import com.opendatagroup.hadrian.errors.PFARuntimeException
import com.opendatagroup.hadrian.jvmcompiler.JavaCode
import com.opendatagroup.hadrian.jvmcompiler.javaSchema
import com.opendatagroup.hadrian.jvmcompiler.javaType
import com.opendatagroup.hadrian.jvmcompiler.JVMNameMangle

import com.opendatagroup.hadrian.ast.AstContext
import com.opendatagroup.hadrian.ast.ExpressionContext
import com.opendatagroup.hadrian.ast.FcnDef
import com.opendatagroup.hadrian.ast.FcnRef

import com.opendatagroup.hadrian.data.PFAArray
import com.opendatagroup.hadrian.data.PFARecord
import com.opendatagroup.hadrian.data.PFAEnumSymbol
import com.opendatagroup.hadrian.data.ComparisonOperator

import com.opendatagroup.hadrian.signature.P
import com.opendatagroup.hadrian.signature.Sig
import com.opendatagroup.hadrian.signature.Signature
import com.opendatagroup.hadrian.signature.Sigs

import com.opendatagroup.hadrian.datatype.Type
import com.opendatagroup.hadrian.datatype.FcnType
import com.opendatagroup.hadrian.datatype.AvroType
import com.opendatagroup.hadrian.datatype.AvroNull
import com.opendatagroup.hadrian.datatype.AvroBoolean
import com.opendatagroup.hadrian.datatype.AvroInt
import com.opendatagroup.hadrian.datatype.AvroLong
import com.opendatagroup.hadrian.datatype.AvroFloat
import com.opendatagroup.hadrian.datatype.AvroDouble
import com.opendatagroup.hadrian.datatype.AvroBytes
import com.opendatagroup.hadrian.datatype.AvroFixed
import com.opendatagroup.hadrian.datatype.AvroString
import com.opendatagroup.hadrian.datatype.AvroEnum
import com.opendatagroup.hadrian.datatype.AvroArray
import com.opendatagroup.hadrian.datatype.AvroMap
import com.opendatagroup.hadrian.datatype.AvroRecord
import com.opendatagroup.hadrian.datatype.AvroField
import com.opendatagroup.hadrian.datatype.AvroUnion

package object tree {
  private var fcns = Map[String, LibFcn]()
  def provides = fcns
  def provide(libFcn: LibFcn): Unit =
    fcns = fcns + Tuple2(libFcn.name, libFcn)

  val prefix = "model.tree."

  //////////////////////////////////////////////////////////////////// 

  object SimpleComparison {
    def apply(datum: PFARecord, field: PFAEnumSymbol, operator: String, value: Any, fieldSchema: Schema, valueSchema: Schema, missingOperators: Boolean): Boolean = {
      if (operator == "alwaysTrue")
        true
      else if (operator == "alwaysFalse")
        false
      else if (missingOperators  &&  operator == "isMissing")
        datum.get(field.value) == null
      else if (missingOperators  &&  operator == "notMissing")
        datum.get(field.value) != null
      else if (operator == "in"  ||  operator == "notIn") {
        if (valueSchema.getType != Schema.Type.ARRAY  ||  checkReaderWriterCompatibility(valueSchema.getElementType, fieldSchema).getType != SchemaCompatibilityType.COMPATIBLE)
          throw new PFARuntimeException("bad value type")
        val vector = value.asInstanceOf[PFAArray[Any]].toVector

        val fieldValue = (datum.get(field.value), valueSchema.getElementType.getType) match {
          case (x: java.lang.Number, Schema.Type.INT) => x.intValue
          case (x: java.lang.Number, Schema.Type.LONG) => x.longValue
          case (x: java.lang.Number, Schema.Type.FLOAT) => x.floatValue
          case (x: java.lang.Number, Schema.Type.DOUBLE) => x.doubleValue
          case (x, _) => x
        }

        val containedInVector = vector.contains(fieldValue)

        if (operator == "in")
          containedInVector
        else
          !containedInVector
      }

      else {
        fieldSchema.getType match {
          case Schema.Type.INT | Schema.Type.LONG | Schema.Type.FLOAT | Schema.Type.DOUBLE => {
            val fieldNumber = datum.get(field.value).asInstanceOf[java.lang.Number].doubleValue
            val valueNumber = value match {
              case null => throw new PFARuntimeException("bad value type")
              case x: Int => x.toDouble
              case x: Long => x.toDouble
              case x: Float => x.toDouble
              case x: Double => x
              case x: java.lang.Number => x.doubleValue
              case _ => throw new PFARuntimeException("bad value type")
            }

            operator match {
              case "<=" => fieldNumber <= valueNumber
              case "<" => fieldNumber < valueNumber
              case ">=" => fieldNumber >= valueNumber
              case ">" => fieldNumber > valueNumber
              case "==" => fieldNumber == valueNumber
              case "!=" => fieldNumber != valueNumber
              case _ => throw new PFARuntimeException("invalid comparison operator")
            }
          }

          case Schema.Type.STRING => {
            val fieldString = datum.get(field.value).asInstanceOf[String]
            val valueString = value match {
              case null => throw new PFARuntimeException("bad value type")
              case x: String => x
              case _ => throw new PFARuntimeException("bad value type")
            }

            operator match {
              case "==" => fieldString == valueString
              case "!=" => fieldString != valueString
              case "<=" => (new ComparisonOperator(valueSchema, -3)).apply(fieldString, valueString)
              case "<" => (new ComparisonOperator(valueSchema, -2)).apply(fieldString, valueString)
              case ">=" => (new ComparisonOperator(valueSchema, 2)).apply(fieldString, valueString)
              case ">" => (new ComparisonOperator(valueSchema, 3)).apply(fieldString, valueString)
              case _ => throw new PFARuntimeException("invalid comparison operator")
            }
          }

          case _ => {
            if (checkReaderWriterCompatibility(valueSchema, fieldSchema).getType != SchemaCompatibilityType.COMPATIBLE)
              throw new PFARuntimeException("bad value type")

            val fieldObject = (datum.get(field.value), valueSchema.getType) match {
              case (x: java.lang.Integer, Schema.Type.INT) => x
              case (x: java.lang.Integer, Schema.Type.LONG) => java.lang.Long.valueOf(x.longValue)
              case (x: java.lang.Integer, Schema.Type.FLOAT) => java.lang.Float.valueOf(x.floatValue)
              case (x: java.lang.Integer, Schema.Type.DOUBLE) => java.lang.Double.valueOf(x.doubleValue)
              case (x: java.lang.Long, Schema.Type.LONG) => x
              case (x: java.lang.Long, Schema.Type.FLOAT) => java.lang.Float.valueOf(x.floatValue)
              case (x: java.lang.Long, Schema.Type.DOUBLE) => java.lang.Double.valueOf(x.doubleValue)
              case (x: java.lang.Float, Schema.Type.FLOAT) => x
              case (x: java.lang.Float, Schema.Type.DOUBLE) => java.lang.Double.valueOf(x.doubleValue)
              case (x: java.lang.Double, Schema.Type.DOUBLE) => x
              case (x, _) => x
            }

            operator match {
              case "<=" => (new ComparisonOperator(valueSchema, -3)).apply(fieldObject, value.asInstanceOf[AnyRef])
              case "<" => (new ComparisonOperator(valueSchema, -2)).apply(fieldObject, value.asInstanceOf[AnyRef])
              case "!=" => (new ComparisonOperator(valueSchema, -1)).apply(fieldObject, value.asInstanceOf[AnyRef])
              case "==" => (new ComparisonOperator(valueSchema, 1)).apply(fieldObject, value.asInstanceOf[AnyRef])
              case ">=" => (new ComparisonOperator(valueSchema, 2)).apply(fieldObject, value.asInstanceOf[AnyRef])
              case ">" => (new ComparisonOperator(valueSchema, 3)).apply(fieldObject, value.asInstanceOf[AnyRef])
              case _ => throw new PFARuntimeException("invalid comparison operator")
            }
          }
        }
      }
    }

    def removeNull(schema: Schema): Schema = schema.getType match {
      case Schema.Type.UNION => {
        val withoutNull = schema.getTypes filter {_.getType != Schema.Type.NULL}
        if (withoutNull.isEmpty)
          schema
        else if (withoutNull.size == 1)
          withoutNull.head
        else
          Schema.createUnion(withoutNull)
      }
      case _ => schema
    }
  }

  ////   simpleTest (SimpleTest)
  object SimpleTest extends LibFcn {
    val name = prefix + "simpleTest"
    val sig = Sig(List("datum" -> P.WildRecord("D", Map()), "comparison" -> P.WildRecord("T", ListMap("field" -> P.EnumFields("F", "D"), "operator" -> P.String, "value" -> P.Wildcard("V")))), P.Boolean)
    val doc =
      <doc>
        <desc>Determine if <p>datum</p> passes a test defined by <p>comparison</p>.</desc>
        <param name="datum">Sample value to test.</param>
        <param name="comparison">Record that describes a test.
          <paramField name="field">Field name from <p>datum</p>: the enumeration type must include all fields of <tp>D</tp> in their declaration order.</paramField>
          <paramField name="operator">One of the following: "==" (equal), "!=" (not equal), "&lt;" (less than), "&lt;=" (less or equal), "&gt;" (greater than), "&gt;=" (greater or equal), "in" (member of a set), "notIn" (not a member of a set), "alwaysTrue" (ignore <pf>value</pf>, return <c>true</c>), "alwaysFalse" (ignore <pf>value</pf>, return <c>false</c>), "isMissing" (ignore <pf>value</pf>, return <c>true</c> iff the field of <p>datum</p> is <c>null</c>), and "notMissing" (ignore <pf>value</pf>, return <c>false</c> iff the field of <p>datum</p> is <c>null</c>).</paramField>
          <paramField name="value">Value to which the field of <p>datum</p> is compared.</paramField>
        </param>
        <ret>Returns <c>true</c> if the field of <p>datum</p> &lt;op&gt; <pf>value</pf> is <c>true</c>, <c>false</c> otherwise, where &lt;op&gt; is the <pf>operator</pf>.</ret>
        <error>Raises an "invalid comparison operator" if <pf>operator</pf> is not one of "==", "!=", "&lt;", "&lt;=", "&gt;", "&gt;=", "in", "notIn", "alwaysTrue", "alwaysFalse", "isMissing", "notMissing".</error>
        <error>Raises a "bad value type" if the <pf>field</pf> of <p>datum</p> and <tp>V</tp> are not both numbers and the <pf>field</pf> cannot be upcast to <tp>V</tp>.</error>
      </doc>

    override def javaCode(args: Seq[JavaCode], argContext: Seq[AstContext], paramTypes: Seq[Type], retType: AvroType): JavaCode =
      JavaCode("""(new Object() {
public boolean apply(com.opendatagroup.hadrian.data.PFARecord datum, %s comparison) {
com.opendatagroup.hadrian.data.PFAEnumSymbol field = (com.opendatagroup.hadrian.data.PFAEnumSymbol)(comparison.%s);
return com.opendatagroup.hadrian.lib1.model.tree.package$SimpleComparison$.MODULE$.apply(datum, field, comparison.%s, comparison.%s, datum.fieldTypes()[field.value()], comparison.fieldTypes()[comparison.fieldIndex("value")], true);
} }).apply((com.opendatagroup.hadrian.data.PFARecord)%s, (%s)%s)""",
        javaType(paramTypes(1).asInstanceOf[AvroRecord], true, true, false),
        JVMNameMangle.s("field"),
        JVMNameMangle.s("operator"),
        JVMNameMangle.s("value"),
        wrapArg(0, args, paramTypes, true),
        javaType(paramTypes(1).asInstanceOf[AvroRecord], true, true, false),
        wrapArg(1, args, paramTypes, true))

  }
  provide(SimpleTest)

  ////   missingTest (MissingTest)
  object MissingTest extends LibFcn {
    val name = prefix + "missingTest"
    val sig = Sig(List("datum" -> P.WildRecord("D", Map()), "comparison" -> P.WildRecord("T", ListMap("field" -> P.EnumFields("F", "D"), "operator" -> P.String, "value" -> P.Wildcard("V")))), P.Union(List(P.Null, P.Boolean)))
    val doc =
      <doc>
        <desc>Determine if <p>datum</p> passes a test defined by <p>comparison</p>, allowing for missing values.</desc>
        <param name="datum">Sample value to test.</param>
        <param name="comparison">Record that describes a test.
          <paramField name="field">Field name from <p>datum</p>: the enumeration type must include all fields of <tp>D</tp> in their declaration order.</paramField>
          <paramField name="operator">One of the following: "==" (equal), "!=" (not equal), "&lt;" (less than), "&lt;=" (less or equal), "&gt;" (greater than), "&gt;=" (greater or equal), "in" (member of a set), "notIn" (not a member of a set), "alwaysTrue" (ignore <pf>value</pf>, return <c>true</c>), "alwaysFalse" (ignore <pf>value</pf>, return <c>false</c>).</paramField>
          <paramField name="value">Value to which the field of <p>datum</p> is compared.</paramField>
        </param>
        <ret>If the field of <p>datum</p> is <c>null</c>, this function returns <c>null</c> (unknown test result).  Otherwise, it returns <p>datum</p> field &lt;op&gt; <pf>value</pf>, where &lt;op&gt; is the <pf>operator</pf></ret>
        <error>Raises an "invalid comparison operator" if <pf>operator</pf> is not one of "==", "!=", "&lt;", "&lt;=", "&gt;", "&gt;=", "in", "notIn", "alwaysTrue", "alwaysFalse".</error>
        <error>Raises a "bad value type" if the <pf>field</pf> of <p>datum</p> and <tp>V</tp> are not both numbers and the <pf>field</pf> cannot be upcast to <tp>V</tp>.</error>
      </doc>

    override def javaCode(args: Seq[JavaCode], argContext: Seq[AstContext], paramTypes: Seq[Type], retType: AvroType): JavaCode =
      JavaCode("""(new Object() {
public java.lang.Boolean apply(com.opendatagroup.hadrian.data.PFARecord datum, %s comparison) {
com.opendatagroup.hadrian.data.PFAEnumSymbol field = (com.opendatagroup.hadrian.data.PFAEnumSymbol)(comparison.%s);
if (datum.get(field.value()) == null)
    return null;
return java.lang.Boolean.valueOf(com.opendatagroup.hadrian.lib1.model.tree.package$SimpleComparison$.MODULE$.apply(datum, field, comparison.%s, comparison.%s, com.opendatagroup.hadrian.lib1.model.tree.package$SimpleComparison$.MODULE$.removeNull(datum.fieldTypes()[field.value()]), comparison.fieldTypes()[comparison.fieldIndex("value")], false));
} }).apply((com.opendatagroup.hadrian.data.PFARecord)%s, (%s)%s)""",
        javaType(paramTypes(1).asInstanceOf[AvroRecord], true, true, false),
        JVMNameMangle.s("field"),
        JVMNameMangle.s("operator"),
        JVMNameMangle.s("value"),
        wrapArg(0, args, paramTypes, true),
        javaType(paramTypes(1).asInstanceOf[AvroRecord], true, true, false),
        wrapArg(1, args, paramTypes, true))
  }
  provide(MissingTest)

  ////   surrogateTest (SurrogateTest)
  object SurrogateTest extends LibFcn {
    val name = prefix + "surrogateTest"
    val sig = Sig(List("datum" -> P.WildRecord("D", Map()), "comparisons" -> P.Array(P.WildRecord("T", Map())), "missingTest" -> P.Fcn(List(P.Wildcard("D"), P.Wildcard("T")), P.Union(List(P.Null, P.Boolean)))), P.Boolean)
    val doc =
      <doc>
        <desc>Apply <p>missingTest</p> to an array of <p>comparisons</p> until one yields a non-null result.</desc>
        <param name="datum">Sample value to test.</param>
        <param name="comparisons">Records that describe the tests.</param>
        <ret>Returns the value of the first test that returns <c>true</c> or <c>false</c>.</ret>
        <error>If all tests return <c>null</c>, this function raises a "no successful surrogate" error.</error>
      </doc>
    def apply(datum: PFARecord, comparisons: PFAArray[PFARecord], missingTest: (PFARecord, PFARecord) => java.lang.Boolean): Boolean = {
      var result = false
      if (comparisons.toVector.indexWhere(comparison =>
        missingTest(datum, comparison) match {
          case null => false
          case x: java.lang.Boolean => {
            result = x.booleanValue
            true
          }
        }) == -1)
        throw new PFARuntimeException("no successful surrogate")
      else
        result
    }
  }
  provide(SurrogateTest)

  ////   simpleWalk (SimpleWalk)
  object SimpleWalk extends LibFcn {
    val name = prefix + "simpleWalk"
    val sig = Sig(List(
      "datum" -> P.WildRecord("D", Map()),
      "treeNode" -> P.WildRecord("T", ListMap("pass" -> P.Union(List(P.WildRecord("T", Map()), P.Wildcard("S"))), "fail" -> P.Union(List(P.WildRecord("T", Map()), P.Wildcard("S"))))),
      "test" -> P.Fcn(List(P.Wildcard("D"), P.Wildcard("T")), P.Boolean)
    ), P.Wildcard("S"))
    val doc =
      <doc>
        <desc>Descend through a tree, testing the fields of <p>datum</p> with the <p>test</p> function using <p>treeNode</p> to define the comparison, continuing to <pf>pass</pf> or <pf>fail</pf> until reaching a leaf node of type <tp>S</tp> (score).</desc>
        <param name="datum">Sample value to test.</param>
        <param name="treeNode">Node of the tree, which contains a predicate to be interpreted by <p>test</p>.
          <paramField name="pass">Branch to follow if <p>test</p> returns <c>true</c>.</paramField>
          <paramField name="fail">Branch to follow if <p>test</p> returns <c>false</c>.</paramField>
        </param>
        <param name="test">Test function that converts <p>datum</p> and <p>treeNode</p> into <c>true</c> or <c>false</c>.</param>
        <ret>Leaf node of type <tp>S</tp>, which must be different from the tree nodes.  For a classification tree, <tp>S</tp> could be a string or an enumeration set.  For a regression tree, <tp>S</tp> would be a numerical type.  For a multivariate regression tree, <tp>S</tp> would be an array of numbers, etc.</ret>
      </doc>
    @tailrec
    def apply(datum: PFARecord, treeNode: PFARecord, test: (PFARecord, PFARecord) => Boolean): AnyRef = {
      val next =
        if (test(datum, treeNode))
          treeNode.get("pass")
        else
          treeNode.get("fail")
      next match {
        case x: PFARecord if (treeNode.getSchema.getFullName == x.getSchema.getFullName) => apply(datum, x, test)
        case x => x
      }
    }
  }
  provide(SimpleWalk)

}

File: titus/titus/lib/model/tree.py
#!/usr/bin/env python

# Copyright (C) 2014  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
# 
# This file is part of Hadrian.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from titus.fcn import Fcn
from titus.fcn import LibFcn
from titus.signature import Sig
from titus.signature import Sigs
from titus.datatype import *
from titus.errors import *
from titus.util import callfcn
import titus.P as P

provides = {}
def provide(fcn):
    provides[fcn.name] = fcn

prefix = "model.tree."

#################################################################### 

def simpleComparison(paramTypes, datum, comparison, missingOperators, parser, code1, code2, fcnName, pos):
    field = comparison["field"]
    fieldValue = datum[field]
    operator = comparison["operator"]
    value = comparison["value"]

    fieldValueType = [x for x in paramTypes[0]["fields"] if x["name"] == field][0]["type"]
    valueType = [x for x in paramTypes[1]["fields"] if x["name"] == "value"][0]["type"]

    if not missingOperators:
        if isinstance(fieldValueType, (list, tuple)):
            withoutNull = [x for x in fieldValueType if x != "null" and x != {"type": "null"}]
            if len(withoutNull) == 1:
                fieldValueType = withoutNull[0]
            else:
                fieldValueType = withoutNull

            if isinstance(fieldValue, dict) and len(fieldValue) == 1 and any(parser.getAvroType(x).name == fieldValue.keys()[0] for x in withoutNull):
                fieldValue, = fieldValue.values()
        
    fieldValueType = parser.getAvroType(fieldValueType)
    valueType = parser.getAvroType(valueType)

    if operator == "alwaysTrue":
        return True
    elif operator == "alwaysFalse":
        return False
    elif missingOperators and operator == "isMissing":
        return (fieldValue is None)
    elif missingOperators and operator == "notMissing":
        return (fieldValue is not None)

    elif operator == "in" or operator == "notIn":
        if isinstance(valueType, AvroArray) and valueType.items.accepts(fieldValueType):
            pass
        elif isinstance(valueType, AvroUnion) and any(isinstance(x, AvroArray) and x.items.accepts(fieldValueType) for x in valueType.types):
            pass
        else:
            raise PFARuntimeException("bad value type", code1, fcnName, pos)

        if isinstance(value, dict) and len(value) == 1 and value.keys() == ["array"]:
            value, = value.values()

        if not missingOperators:
            if fieldValue is None:
                return None

        containedInList = (fieldValue in value)
        if operator == "in":
            return containedInList
        else:
            return not containedInList

    else:
        if not valueType.accepts(fieldValueType):
            if isinstance(valueType, (AvroInt, AvroLong, AvroFloat, AvroDouble)) and \
               isinstance(fieldValueType, (AvroInt, AvroLong, AvroFloat, AvroDouble)):
                pass
            else:
                raise PFARuntimeException("bad value type", code1, fcnName, pos)

        if not missingOperators and fieldValue is None:
            return None

        if isinstance(fieldValueType, (AvroInt, AvroLong, AvroFloat, AvroDouble)):
            if isinstance(value, dict) and (value.keys() == ["int"] or value.keys() == ["long"] or value.keys() == ["float"] or value.keys() == ["double"]):
                value, = value.values()
            if not isinstance(value, (int, long, float)):
                raise PFARuntimeException("bad value type", code1, fcnName, pos)

            if operator == "<=":
                return fieldValue <= value
            elif operator == "<":
                return fieldValue < value
            elif operator == ">=":
                return fieldValue >= value
            elif operator == ">":
                return fieldValue > value
            elif operator == "==":
                return fieldValue == value
            elif operator == "!=":
                return fieldValue != value
            else:
                raise PFARuntimeException("invalid comparison operator", code2, fcnName, pos)

        else:
            if operator == "<=":
                return compare(valueType, fieldValue, value) <= 0
            elif operator == "<":
                return compare(valueType, fieldValue, value) < 0
            elif operator == ">=":
                return compare(valueType, fieldValue, value) >= 0
            elif operator == ">":
                return compare(valueType, fieldValue, value) > 0
            elif operator == "==":
                return compare(valueType, fieldValue, value) == 0
            elif operator == "!=":
                return compare(valueType, fieldValue, value) != 0
            else:
                raise PFARuntimeException("invalid comparison operator", code2, fcnName, pos)

class SimpleTest(LibFcn):
    name = prefix + "simpleTest"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"comparison": P.WildRecord("T", {"field": P.EnumFields("F", "D"), "operator": P.String(), "value": P.Wildcard("V")})}], P.Boolean())
    errcodeBase = 32000
    def __call__(self, state, scope, pos, paramTypes, datum, comparison):
        return simpleComparison(paramTypes, datum, comparison, True, state.parser, self.errcodeBase + 1, self.errcodeBase + 0, self.name, pos)
provide(SimpleTest())

class CompoundTest(LibFcn):
    name = prefix + "compoundTest"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"operator": P.String()}, {"comparisons": P.Array(P.WildRecord("T", {}))}, {"test": P.Fcn([P.Wildcard("D"), P.Wildcard("T")], P.Boolean())}], P.Boolean())
    errcodeBase = 32020
    def __call__(self, state, scope, pos, paramTypes, datum, operator, comparisons, test):
        if operator == "and":
            for comparison in comparisons:
                if callfcn(state, scope, test, [datum, comparison]) is False:
                    return False
            return True
        elif operator == "or":
            for comparison in comparisons:
                if callfcn(state, scope, test, [datum, comparison]) is True:
                    return True
            return False
        elif operator == "xor":
            numTrue = 0
            for comparison in comparisons:
                if callfcn(state, scope, test, [datum, comparison]) is True:
                    numTrue += 1
            return numTrue % 2 == 1
        else:
            raise PFARuntimeException("unrecognized logical operator", self.errcodeBase + 0, self.name, pos)
provide(CompoundTest())

class MissingTest(LibFcn):
    name = prefix + "missingTest"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"comparison": P.WildRecord("T", {"field": P.EnumFields("F", "D"), "operator": P.String(), "value": P.Wildcard("V")})}], P.Union([P.Null(), P.Boolean()]))
    errcodeBase = 32010
    def __call__(self, state, scope, pos, paramTypes, datum, comparison):
#         newDatumTypeFields = [{"name": x["name"], "type": removeNull(x["type"])} if x["name"] == comparison["field"] else x for x in paramTypes[0]["fields"]]
#         newParamTypes = [dict(paramTypes[0], fields=newDatumTypeFields)] + paramTypes[1:]
#         return simpleComparison(newParamTypes, datum, comparison, False, state.parser)
        out = simpleComparison(paramTypes, datum, comparison, False, state.parser, self.errcodeBase + 1, self.errcodeBase + 0, self.name, pos)
        if out is True:
            return {"boolean": True}
        elif out is False:
            return {"boolean": False}
        else:
            return out
provide(MissingTest())

class SurrogateTest(LibFcn):
    name = prefix + "surrogateTest"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"comparisons": P.Array(P.WildRecord("T", {}))}, {"missingTest": P.Fcn([P.Wildcard("D"), P.Wildcard("T")], P.Union([P.Null(), P.Boolean()]))}], P.Boolean())
    errcodeBase = 32030
    def __call__(self, state, scope, pos, paramTypes, datum, comparisons, missingTest):
        for comparison in comparisons:
            result = callfcn(state, scope, missingTest, [datum, comparison])
            if result is not None:
                if isinstance(result, dict) and result.keys() == ["boolean"]:
                    return result.values()[0]
                else:
                    return result
        raise PFARuntimeException("no successful surrogate", self.errcodeBase + 0, self.name, pos)
provide(SurrogateTest())

class SimpleWalk(LibFcn):
    name = prefix + "simpleWalk"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"treeNode": P.WildRecord("T", {"pass": P.Union([P.WildRecord("T", {}), P.Wildcard("S")]), "fail": P.Union([P.WildRecord("T", {}), P.Wildcard("S")])})}, {"test": P.Fcn([P.WildRecord("D", {}), P.WildRecord("T", {})], P.Boolean())}], P.Wildcard("S"))
    errcodeBase = 32040
    def __call__(self, state, scope, pos, paramTypes, datum, treeNode, test):
        treeNodeTypeName = paramTypes[1]["name"]
        node = treeNode
        while True:
            if callfcn(state, scope, test, [datum, node]):
                union = node["pass"]
            else:
                union = node["fail"]
            if union is None:
                node = None
                break
            else:
                (utype, node), = union.items()
                if utype != treeNodeTypeName:
                    break
        return node
provide(SimpleWalk())

class MissingWalk(LibFcn):
    name = prefix + "missingWalk"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"treeNode": P.WildRecord("T", {"pass": P.Union([P.WildRecord("T", {}), P.Wildcard("S")]), "fail": P.Union([P.WildRecord("T", {}), P.Wildcard("S")]), "missing": P.Union([P.WildRecord("T", {}), P.Wildcard("S")])})}, {"test": P.Fcn([P.WildRecord("D", {}), P.WildRecord("T", {})], P.Union([P.Null(), P.Boolean()]))}], P.Wildcard("S"))
    errcodeBase = 32050
    def __call__(self, state, scope, pos, paramTypes, datum, treeNode, test):
        treeNodeTypeName = paramTypes[1]["name"]
        node = treeNode
        while True:
            result = callfcn(state, scope, test, [datum, node])
            if result is True or result == {"boolean": True}:
                union = node["pass"]
            elif result is False or result == {"boolean": False}:
                union = node["fail"]
            elif result is None:
                union = node["missing"]
            if union is None:
                node = None
                break
            else:
                (utype, node), = union.items()
                if utype != treeNodeTypeName:
                    break
        return node
provide(MissingWalk())

class SimpleTree(LibFcn):
    name = prefix + "simpleTree"
    sig = Sig([{"datum": P.WildRecord("D", {})}, {"treeNode": P.WildRecord("T", {"field": P.EnumFields("F", "D"), "operator": P.String(), "value": P.Wildcard("V"), "pass": P.Union([P.WildRecord("T", {}), P.Wildcard("S")]), "fail": P.Union([P.WildRecord("T", {}), P.Wildcard("S")])})}], P.Wildcard("S"))
    errcodeBase = 32060
    def __call__(self, state, scope, pos, paramTypes, datum, treeNode):
        treeNodeTypeName = paramTypes[1]["name"]
        node = treeNode
        while True:
            if simpleComparison(paramTypes, datum, node, True, state.parser, self.errcodeBase + 1, self.errcodeBase + 0, self.name, pos):
                union = node["pass"]
            else:
                union = node["fail"]
            if union is None:
                node = None
                break
            else:
                (utype, node), = union.items()
                if utype != treeNodeTypeName:
                    break
        return node
provide(SimpleTree())
